### DISCUSSION TITLE
Attiny 3226 how to use two analogRead ? No reading

### DISCUSSION THREAD

**Discussion by Brendy1986** [SpenceKonde/megaTinyCore#865] [Q&A]
HI
I use the first analog (A3)  pin PA3 to read the battery voltage and I have declared these lines in setup 
```
  ADC0.CTRLA = 1 << ADC_ENABLE_bm & ~(1 << ADC_RUNSTDBY_bm); 
  ADC0.CTRLC = ADC_REFSEL_2500MV_gc; 
  ADC0.CTRLA = (1 << ADC_ENABLE_bp) & (~(1 << ADC_RUNSTDBY_bp) & ~(1 << ADC_LOWLAT_bp)); 
```
when I wanted to add a second analog to read the value from the photoresistor, I still have the same value because these lines in setup make a collision.
How to solve it so that it can read the voltage using these 3 lines and read the photoresistor?

regards Brendy

**Comment by mosqu-ito** [comment]
If you feel the need to work the registers yourself, you will also need to set the input multiplexor to select a pin as per page 432 of the datasheet.  But most people just let the core handle all that just like with any other Arduino ie. analogRead(PIN_PA3) etc. as described here:
https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Ref_Analog.md

**Comment by SpenceKonde** [comment]
Where did the fuicking response I wrote last night go?!

There is no reason to touch any of those registers.  In fact, the moment you touch any registers except through the API calls clearly described in the analog reference (https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Ref_Analog.md), ALL FUNCTIONALITY RELATED TO THAT PERIPHERAL IS UNDEFINED. analogReturn might work, it might product bogus values, or it might cause the chip to explode - the core provides no guarantees or specification about what to expect (and if something made the chips explode, they would just note that the use must ensure that that situation is not produced). 

Manual vs core-mediated control of a peripheral is either or. You may EITHER use the API functions, and not call ANY of the related API functions and assume full responsibility (do if you change any register under ADC0, you must not expect correct results from analogRead, analogReadEnh or correct behavior from any othher ADC-related functionality. The same holds true of (except where specifically noted in the core reference documentation) for EVERY peripheral. It's not possible to reinitialize every peripheral, as that is a much bigger routine than init_someperiph(), as it also has to zero out every unused register, ad account for any enable locking and configuration change protection. On some of the megatinycore parts, if i did this, the largest flash size available in that pincount could not fit all of the "prior-manual-configuration-proof" initialization code.

And the hardware has no facility to exclude or compensate for the possibility that registers have been modified since initialization - that simillarly breaks all expectation of correct behavior. 

This is clearly stated in the Readme.md - if you're modifying the registers outside of functions we provide for that purpose, any correct behavior is luck - except as noted for limited cases for specific reasons, the nature of which is described at that point inj the docs.

What is truly baffling about your question is that **THERE IS NO REASON TO MANUALLY TOUCH REGISTERS AT ALL**
```c++
ADCPowerOptions(ENABLE_ON | LOW_LAT_OFF);
analogReference(INTERNAL2V5);
Serial.println(analogRead(mypin))';
```

Reading a battery? How is it driving the chip? Usually when i care about power consumption, I run the chip straight off a LiPo battery (Are you using a boost converter or something)? 

in thecase where the chip runs straight from battery it is trivial:
analogReference(INTERNAL1024);
val =analogRead(ADC_VDDDIV10); 
If you had 3.7V battery, divideby 10 gives 370mV
Hence, val will come out as 1024/1000 * (Vdd / 10). For greater precision use analogReadResolution kick analogRead up to 12 bit, and analogReadEnh() (detailed in above linked reference) to use oversampling and decimation to get readings with a "claimed" resolution of 17 bits. I do''t know how many of the low ones are noise under hobbyist conditions.

There are also features to change the sampling duration for high imopedance sources,  perform *true* differential ADC measurements (not the fake ons the Dx has), and use the programmable gain amplifier - EVEN IN SINGLE ENDED MODE!

**Reply by ghost** [reply]
I don't know where to post my question! hope you don't mind if I post it here!

I need to add some lines to my interrupt pins file so I don't get this error:
In file included from C:\Users\3DPRIN~1\AppData\Local\Temp\arduino_build_367455\sketch\NewEncoder.h:19:0,
                 from C:\Users\3d print\Desktop\stuff\NewEncoder-master\examples\SingleEncoder\SingleEncoder.ino:2:
C:\Users\3DPRIN~1\AppData\Local\Temp\arduino_build_367455\sketch\interrupt_pins.h:326:2: error: #error "Interrupts are unknown for this board, please add to this code"
 #error "Interrupts are unknown for this board, please add to this code"
  ^~~~~
C:\Users\3DPRIN~1\AppData\Local\Temp\arduino_build_367455\sketch\interrupt_pins.h:329:2: error: #error "Encoder requires interrupt pins, but this board does not have any :("
 #error "Encoder requires interrupt pins, but this board does not have any :("
  ^~~~~
C:\Users\3DPRIN~1\AppData\Local\Temp\arduino_build_367455\sketch\interrupt_pins.h:330:2: error: #error "You could try defining ENCODER_DO_NOT_USE_INTERRUPTS as a kludge."
 #error "You could try defining ENCODER_DO_NOT_USE_INTERRUPTS as a kludge."
  ^~~~~
exit status 1
Error compiling for board ATtiny3226/3216/1626/1616/1606/826/816/806/426/416/406.

could you point me in the right direction? I would be ever so grateful! really! THANKS in advance!

**Comment by Brendy1986** [comment]
Spence thanks for the detailed answer. Lots of interesting information to read. Thanks for the links.
I've made many circuits on the Attiny Series 2 and love them, among other things, for the very accurate measurement of battery voltage.
The chip is battery powered and measures the battery voltage 4.2-3.3V and goes to sleep. I have done many tests and the effects are very good according to me very accurately.
I use only the processor to measure the battery. I only calibrate the divider once to the battery voltage. I use a rhetoric divider, when the processor wakes up, I turn on the ground on the divider with a pin, it does the measurement and before going to sleep, the divider is disconnected thanks to which it does not consume only 0.8 uA from rtc during sleep.
I will have to try your lines.
![adc](https://user-images.githubusercontent.com/118827809/203643738-63835d9a-3556-40a5-9943-6d031bd522da.png)

  _

> There are also features to change the sampling duration for high imopedance sources, perform true differential ADC measurements (not the fake ons the Dx has), and use the programmable gain amplifier - EVEN IN SINGLE ENDED MODE!

_

Are there any code examples?

**Comment by SpenceKonde** [comment]
Oh, uh if you're reading VDD, why not use use the smallest reference possible 1.024V, and analogRead(VDDDIV10) input? 

with analogReadResolution(12); analogReference(INTERNAL1V024); analogRead(VDDIODIV10), that would mean reading of 4095 would be full scale, and would indicate that Vdd wat at 10.24v, and that the chips is likely on fire. More realistic values might be like 2000.
